expr/expr.go

```diff
@@ -3266,6 +3267,82 @@ func EvalExpr(e *expr, from, until int32, values map[MetricRequest][]*MetricData
 		}
 
 		return results, nil
+
+	case "cactiStyle": // cactiStyle(seriesList, system=None, units=None)
+		// Get the series data
+		original, err := getSeriesArg(e.args[0], from, until, values)
+		if err != nil {
+			return nil, err
+		}
+
+		// Get the arguments
+		system, err := getStringNamedOrPosArgDefault(e, "system", 1, "")
+		if err != nil {
+			return nil, err
+		}
+		unit, err := getStringNamedOrPosArgDefault(e, "units", 2, "")
+		if err != nil {
+			return nil, err
+		}
+
+		// Deal with each of the series
+		for _, a := range original {
+			// Calculate min, max, current
+			//
+			// This saves calling summarizeValues 3 times and looping over
+			// the metrics 3 times
+			//
+			// For min:
+			// Ignoring any absent values and inf (if we have a value)
+			// Using summarizeValues("min", ...) results in incorrect values, when absent
+			// values are present
+			//
+			val_min := math.Inf(-1)
+			val_current := math.Inf(-1)
+			val_max := math.Inf(-1)
+			for i, av := range a.GetValues() {
+				if !a.IsAbsent[i] && (val_min == math.Inf(-1) || av < val_min) {
+					val_min = av
+				}
+				if !a.IsAbsent[i] && av > val_max {
+					val_max = av
+				}
+				val_current = av
+			}
+
+			// Format the output correctly
+			min := ""
+			max := ""
+			current := ""
+			if system == "si" {
+				mv, mf := humanize.ComputeSI(val_min)
+				xv, xf := humanize.ComputeSI(val_max)
+				cv, cf := humanize.ComputeSI(val_current)
+
+				min = fmt.Sprintf("%.0f%s", mv, mf)
+				max = fmt.Sprintf("%.0f%s", xv, xf)
+				current = fmt.Sprintf("%.0f%s", cv, cf)
+
+			} else if system == "" {
+				min = fmt.Sprintf("%.0f", val_min)
+				max = fmt.Sprintf("%.0f", val_max)
+				current = fmt.Sprintf("%.0f", val_current)
+
+			} else {
+				return nil, fmt.Errorf("%s is not supported for system", system)
+			}
+
+			// Append the unit if specified
+			if len(unit) > 0 {
+				min = fmt.Sprintf("%s %s", min, unit)
+				max = fmt.Sprintf("%s %s", max, unit)
+				current = fmt.Sprintf("%s %s", current, unit)
+			}
+
+			a.Name = proto.String(fmt.Sprintf("%s Current: %s Max: %s Min: %s", *a.Name, current, max, min))
```

Comment:
We don't mangle the name in-place (it messes up the tests for one).
Instead follow our convention of:

```go
    //above the for loop
    var results []*MetricData

    for _, a := range original {
        ...
    
        r := *a
        r.Name = proto.String(fmt.Sprintf("%s Current: %s Max: %s Min: %s", *a.Name, current, max, min))
        results = append(results, &r)
    }
    return results, nil
```